File: sample/src/main/java/com/fondesa/recyclerviewdivider/sample/MainActivity.java
package com.fondesa.recyclerviewdivider.sample;

import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.fondesa.recyclerviewdivider.RecyclerViewDivider;
import com.fondesa.recyclerviewdivider.factories.MarginFactory;
import com.fondesa.recyclerviewdivider.factories.SizeFactory;
import com.fondesa.recyclerviewdivider.factories.TintFactory;
import com.fondesa.recyclerviewdivider.factories.VisibilityFactory;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {
    private static final int LIST_SIZE = 18;
    private static final int SPAN_COUNT = 2;
    private static final String SHOW = "ALL";
    private static final String REMOVE = "REMOVE";
    boolean dividerShown;

    private RecyclerViewDivider firstDivider;
    private RecyclerViewDivider secondDivider;

    @SuppressWarnings("ConstantConditions")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        RecyclerView mFirstRecyclerView = (RecyclerView) findViewById(R.id.first_recycler_view);
        ((GridLayoutManager) mFirstRecyclerView.getLayoutManager()).setSpanCount(SPAN_COUNT);

        RecyclerView mSecondRecyclerView = (RecyclerView) findViewById(R.id.second_recycler_view);
        ((GridLayoutManager) mSecondRecyclerView.getLayoutManager()).setSpanCount(SPAN_COUNT);

        firstDivider = RecyclerViewDivider.with(this)
                .addTo(mFirstRecyclerView)
                .visibilityFactory(new VisibilityFactory() {
                    @Override
                    public int displayDividerForItem(int groupCount, int groupIndex) {
                        if (groupIndex == groupCount - 1) {
                            return SHOW_ITEMS_ONLY;
                        }
                        return SHOW_ALL;
                    }
                })
//                .sizeFactory(new SizeFactory() {
//                    @Override
//                    public int sizeForItem(@Nullable Drawable drawable, int orientation, int listSize, int position) {
//                        return position % 2 == 0 ? 40 : 10;
//                    }
//                })
//                .tintFactory(new TintFactory() {
//                    @Override
//                    public int tintForItem(int listSize, int position) {
//                        return position % 3 == 0 ? Color.RED : Color.GREEN;
//                    }
//                })
                .color(Color.RED)
//                .size(getResources().getDimensionPixelSize(R.dimen.first_div_size))
                .size(20)
//                .marginSize(20)
                .build();

        firstDivider.attach();

        secondDivider = RecyclerViewDivider.with(this)
                .addTo(mSecondRecyclerView)
                .visibilityFactory(new VisibilityFactory() {
                    @Override
                    public int displayDividerForItem(int groupCount, int groupIndex) {
//                        if (groupIndex == groupCount - 1) {
//                            return SHOW_ITEMS_ONLY;
//                        }
                        return SHOW_ALL;
                    }
                })
//                .sizeFactory(new SizeFactory() {
//                    @Override
//                    public int sizeForItem(@Nullable Drawable drawable, int orientation, int listSize, int position) {
//                        return position % 2 == 0 ? 40 : 10;
//                    }
//                })
//                .tintFactory(new TintFactory() {
//                    @Override
//                    public int tintForItem(int listSize, int position) {
//                        return position % 3 == 0 ? Color.RED : Color.GREEN;
//                    }
//                })
                .color(Color.RED)
                .size(20)
                .marginSize(20)
                .build();

        secondDivider.attach();

        dividerShown = true;

        final List<Integer> dummyList = new ArrayList<>(LIST_SIZE);
        for (int i = 0; i < LIST_SIZE; i++) {
            dummyList.add(i + 1);
        }
        final DummyAdapter firstDummyAdapter = new DummyAdapter(true);
        mFirstRecyclerView.setAdapter(firstDummyAdapter);
        final DummyAdapter secondDummyAdapter = new DummyAdapter(false);
        mSecondRecyclerView.setAdapter(secondDummyAdapter);

        firstDummyAdapter.updateList(dummyList);
        secondDummyAdapter.updateList(dummyList);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        super.onCreateOptionsMenu(menu);
        getMenuInflater().inflate(R.menu.main, menu);
        return true;
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        super.onPrepareOptionsMenu(menu);
        menu.findItem(R.id.action_toggle_div).setTitle(dividerShown ? REMOVE : SHOW);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        super.onOptionsItemSelected(item);
        if (item.getItemId() == R.id.action_toggle_div) {
            if (dividerShown) {
                firstDivider.detach();
                secondDivider.detach();
            } else {
                firstDivider.attach();
                secondDivider.attach();
            }
            dividerShown = !dividerShown;
            invalidateOptionsMenu();
        }
        return true;
    }

    private class DummyAdapter extends RecyclerView.Adapter {
        private List<Integer> list;
        private boolean first;

        public DummyAdapter(boolean first) {
            this.list = new ArrayList<>();
            this.first = first;
        }

        @Override
        public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            return new DummyViewHolder(LayoutInflater.from(MainActivity.this).inflate(first ? R.layout.dummy_cell_first : R.layout.dummy_cell_second, parent, false));
        }

        @Override
        public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
            ((DummyViewHolder) holder).setItemText(list.get(position));
        }

        @Override
        public int getItemCount() {
            return list.size();
        }

        public void updateList(List<Integer> list) {
            if (list != this.list) {
                this.list = list;
                notifyDataSetChanged();
            }
        }
    }

    private static class DummyViewHolder extends RecyclerView.ViewHolder {
        public DummyViewHolder(View itemView) {
            super(itemView);
        }

        public void setItemText(int item) {
            ((TextView) itemView).setText(String.valueOf(item));
        }
    }
}

File: sample/src/main/kotlin/com/fondesa/recyclerviewdivider/sample/MainActivity.kt
/*
 * Copyright (c) 2020 Fondesa
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.fondesa.recyclerviewdivider.sample

import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.fondesa.recyclerviewdivider.RecyclerViewDivider
import java.util.ArrayList

class MainActivity : AppCompatActivity() {

    companion object {
        private const val LIST_SIZE = 16
        private const val SPAN_COUNT = 3
        private const val SHOW = "ALL"
        private const val REMOVE = "REMOVE"
    }

    private var dividerShown: Boolean = false

    private lateinit var firstDivider: RecyclerViewDivider
    private lateinit var secondDivider: RecyclerViewDivider

    private val firstRecyclerView by lazy { findViewById<RecyclerView>(R.id.first_recycler_view) }
    private val secondRecyclerView by lazy { findViewById<RecyclerView>(R.id.second_recycler_view) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val firstManager = firstRecyclerView.layoutManager as GridLayoutManager
        firstManager.spanCount = SPAN_COUNT
        firstManager.spanSizeLookup = DummyLookup()

        val secondManager = secondRecyclerView.layoutManager as GridLayoutManager
        secondManager.spanCount = 1
//        secondManager.spanSizeLookup = DummyLookup()

        firstDivider = RecyclerViewDivider.with(this)
            .size(24)
            .build()

        firstDivider.addTo(firstRecyclerView)

        secondDivider = RecyclerViewDivider.with(this)
            .color(Color.BLACK)
            .inset(0, 70)
            .build()

        secondDivider.addTo(secondRecyclerView)

        dividerShown = true

        val dummyList = (0 until LIST_SIZE).map { it + 1 }.toList()
        val firstDummyAdapter = DummyAdapter(true)
        firstRecyclerView.adapter = firstDummyAdapter
        val secondDummyAdapter = DummyAdapter(false)
        secondRecyclerView.adapter = secondDummyAdapter

        firstDummyAdapter.updateList(dummyList)
        secondDummyAdapter.updateList(dummyList)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        super.onCreateOptionsMenu(menu)
        menuInflater.inflate(R.menu.main, menu)
        return true
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        super.onPrepareOptionsMenu(menu)
        menu.findItem(R.id.action_toggle_div).title = if (dividerShown) REMOVE else SHOW
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        super.onOptionsItemSelected(item)
        if (item.itemId == R.id.action_toggle_div) {
            if (dividerShown) {
                firstDivider.removeFrom(firstRecyclerView)
                secondDivider.removeFrom(secondRecyclerView)
            } else {
                firstDivider.addTo(firstRecyclerView)
                secondDivider.addTo(secondRecyclerView)
            }
            dividerShown = !dividerShown
            invalidateOptionsMenu()
        }
        return true
    }

    private inner class DummyAdapter internal constructor(private val first: Boolean) :
        RecyclerView.Adapter<DummyViewHolder>() {
        private var list: List<Int>? = null

        init {
            this.list = ArrayList()
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DummyViewHolder =
            DummyViewHolder(
                LayoutInflater.from(this@MainActivity)
                    .inflate(
                        if (first) R.layout.dummy_cell_first else R.layout.dummy_cell_second,
                        parent,
                        false
                    )
            )

        override fun onBindViewHolder(holder: DummyViewHolder, position: Int) {
            holder.setItemText(list!![position])
        }

        override fun getItemCount(): Int = list!!.size

        internal fun updateList(list: List<Int>) {
            if (list !== this.list) {
                this.list = list
                notifyDataSetChanged()
            }
        }
    }

    private class DummyViewHolder internal constructor(itemView: View) :
        RecyclerView.ViewHolder(itemView) {

        internal fun setItemText(item: Int) {
            (itemView as TextView).text = item.toString()
        }
    }

    private class DummyLookup : GridLayoutManager.SpanSizeLookup() {

        override fun getSpanSize(position: Int): Int = position % 3 + 1
    }
}
